File: TicTacToe.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <list>
#include <algorithm>
#include <vector>
#include <random>

using namespace std;

unsigned int randomInt(unsigned int min, unsigned int max)
{
    int r;
    const unsigned int range = 1 + max - min;
    const unsigned int buckets = RAND_MAX / range;
    const unsigned int limit = buckets * range;

    /* Create equal size buckets all in a row, then fire randomly towards
     * the buckets until you land in one of them. All buckets are equally
     * likely. If you land off the end of the line of buckets, try again. */
    do
    {
        r = rand();
    } while (r >= limit);

    return min + (r / buckets);
}

class TicTacToe {
    private:
        int game[3][3];
        int turn;
        bool plot;
        int hasFinished;
        int seed;

        void printGame()
        {
            cout << "\n ___  ___  ___ \n";
            for (int i=0; i<3; i++) {
                for (int j=0; j<3; j++) {
                    if (game[i][j] == 0) {
                        cout << "|   |";
                    }
                    if (game[i][j] == 1) {
                        cout << "| x |";
                    }
                    if (game[i][j] == 2) {
                        cout << "| o |";
                    }
                }
                cout << "\n|___||___||___|";
                cout << "\n";
            }
        }
        int checkWinner(int gameTile)
        {
            if (gameTile == 1) {
                if (plot == true) {
                    cout << "\n 'X' wins!\n";
                }
                return -1;
            }
            if (gameTile == 2) {
                if (plot == true) {
                    cout << "\n 'O' wins!\n";
                }
                return 1;
            }
        }
        int checkWin()
        {
            if ((game[0][0] == game[1][1]) && (game[1][1] == game[2][2]) && (game[0][0] != 0)) {
                return checkWinner(game[0][0]);
            }
            if ((game[0][2] == game[1][1]) && (game[1][1] == game[2][0]) && (game[0][2] != 0)) {
                return checkWinner(game[0][2]);
            }
            for (int i=0; i<3; i++) {
                if ((game[0][i] == game[1][i]) && (game[1][i] == game[2][i]) && (game[0][i] != 0))
                    return checkWinner(game[0][i]);
                if ((game[i][0] == game[i][1]) && (game[i][1] == game[i][2]) && (game[i][0] != 0))
                    return checkWinner(game[i][0]);
            }
            for (int i=0; i<3; i++) {
                for (int j=0; j<3; j++) {
                    if (game[i][j] == 0) {
                        return 2;
                    }
                }
            }
            if (plot == true) {
                cout << "\nDraw!";
            }
            return 0;
        }
        void playerInput() {
            int x;
            int y;
            while (true) {
                cout << "Input x coordinate: " << endl;
                cin >> x;
                cout << "Input y coordinate: " << endl;
                cin >> y;
                if ((x > 2) || (x < 0) || (y < 0) || (y > 2)) {
                    cout << "Invalid coordinates! Please enter a value of x and y between 0 and 2." << endl;
                }
                if (game[x][y] != 0) {
                    cout << "Tile already filled! Please enter another coordinate." << endl;
                }
                else {
                    game[x][y] = 1;
                    break;
                }
            }
        }
        void computerInput() {
            int x;
            int y;
            while (true) {
                x = randomInt(0, 2);
                y = randomInt(0, 2);
                if (game[x][y] == 0) {
                    game[x][y] = 2;
                    break;
                }
            }
        }
    public:
        TicTacToe () {
            seed = randomInt(0, 1);
            if (plot == true) {
                if (seed == 0) {
                    cout << "'X' begins the game!" << endl;
                }
                else {
                    cout << "'O' begins the game!" << endl;
                }
            }
        }
        void setValue(int initialGame[3][3], int initialTurn, bool plotGame)
        {
            memcpy(game, initialGame, 9*sizeof(int));
            turn = initialTurn;
            plot = plotGame;
        }
        void runGame() {
            if (plot == true) {
                cout << "\nBeginning game ..." << endl;
            }
            hasFinished = checkWin();
            while (hasFinished == 2) {
                if (plot == true) {
                    cout << "\nTurn: " << turn << endl;
                    cout << "\n------------------------------------\n" << endl;
                }
                if (seed == 0) {
                    playerInput();
                    hasFinished = checkWin();
                    if (hasFinished == 2) {
                        computerInput();
                        hasFinished = checkWin();
                        if (plot == true) {
                            printGame();
                        }
                    }
                    else {
                        if (plot == true) {
                            printGame();
                        }
                    }
                }
                else {
                    computerInput();
                    if (plot == true) {
                        printGame();
                    }
                    hasFinished = checkWin();
                    if (hasFinished == 2) {
                        playerInput();
                        hasFinished = checkWin();
                        if (plot == true) {
                            printGame();
                        }
                    }
                    else {
                        if (plot == true) {
                            printGame();
                        }
                    }
                    if (plot == true) {
                    cout << "------------------------------------\n" << endl;
                    }
                }
                turn++;
            }
        }
        ~TicTacToe () { }
};

int main(void) {
    TicTacToe game1;
    int initialGame[3][3] = {0, 0, 0, 0, 0, 0, 0, 0, 0};
    while (true) {
        game1.setValue(initialGame, 0, true);
        game1.runGame();
    }
    
    return 0;
}

File: python/TicTacToe.py
import numpy as np
import copy
import random

class TicTacToe:
    def __init__(self, initialGame=None, initialTurn=None, plotGame=None):
        self._empty = [["_" for i in range(3)] for j in range(3)]
        if initialGame is None:
            initialGame = copy.copy(self._empty)
        if initialTurn is None:
            initialTurn = 0
        if plotGame is None:
            plotGame = False

        self._game = copy.copy(initialGame)
        self._turn = initialTurn
        self._plot = plotGame
        self._has_finished = self.check_win()
        self._seed = random.randint(0, 1) # random seed to check who begins

        if (self._plot):
            if self._seed == 0:
                print("'X' begins the game!")
            else:
                print("'O' begins the game!")

    # Print game
    def print_game(self):
        print("\n-------------------------------------\n")
        print(np.matrix(self._game))
        print("\n-------------------------------------\n")

    # Check who is winner
    def check_winner(self, game_tile):
        if game_tile == "x":
            if (self._plot):
                print("'X' wins!")
            return -1
        if game_tile == "o":
            if (self._plot):
                print("'O' wins!")
            return 1

    # Check if there is a winner
    def check_win(self):
        # Check diagonals
        if ((self._game[0][0] == self._game[1][1]) and (self._game[1][1] == self._game[2][2]) and (self._game[0][0] != "")):
            return self.check_winner(self._game[0][0])
        if ((self._game[0][2] == self._game[1][1]) and (self._game[1][1] == self._game[2][0]) and (self._game[0][2] != "")):
            return self.check_winner(self._game[0][2])
        # Check straight lines
        for i in range(3):
            if ((self._game[0][i] == self._game[1][i]) and (self._game[1][i] == self._game[2][i]) and (self._game[0][i] != "")):
                return self.check_winner(self._game[0][i])
            if ((self._game[i][0] == self._game[i][1]) and (self._game[i][1] == self._game[i][2]) and (self._game[i][0] != "")):
                return self.check_winner(self._game[i][0])
        # Check for draw
        for i in range(3):
            for j in range(3):
                if self._game[i][j] == "_":
                    return None
        if (self._game):
            print("Draw!")
        return 0

    def player1_input(self):
        pass

    def player2_input(self):
        pass

    def run_game(self):
        # Loops game until victory or draw
        while (self._has_finished == None):
            print("Turn: {}".format(self._turn))
            if self._seed == 0: # player 1 begins the game
                self.player1_input()
                self._has_finished = self.check_win()
                if self._has_finished == None:
                    self.player2_input()
                    self._has_finished = self.check_win()
                    self.print_game()
                else:
                    self.print_game()
            else: # player 2 begins the game
                self.player2_input()
                self.print_game()
                self._has_finished = self.check_win()
                if self._has_finished == None:
                    self.player1_input()
                    self._has_finished = self.check_win()
                    self.print_game()
                else:
                    self.print_game()
            self._turn += 1
    
    def clear_game(self):
        self._game = self._empty

class HumanVsRandom(TicTacToe):   
    def player1_input(self):
        while True:
            try:
                x = int(input("Input x coordinate: "))
                y = int(input("Input y coordinate: "))
                if ((x > 2) or (x < 0) or (y < 0) or (y > 2)):
                    print("Invalid coordinate! Please enter a value of x and y between 0 and 2.")
                elif (self._game[x][y] != "_"):
                    print("Tile already filled! Please enter another coordinate.")
                else:
                    self._game[x][y] = 'x'
                    break
            except ValueError:
                print("The coordinates you entered are not valid! Please enter valid coordinates.")

    def player2_input(self):
        while True:
            x = random.randint(0, 2)
            y = random.randint(0, 2)
            if self._game[x][y] == "_": # Check if tile is empty
                self._game[x][y] = 'o'
                break

class ComputerVsComputer(TicTacToe):
    def player1_input(self):
        while True:
            x = random.randint(0, 2)
            y = random.randint(0, 2)
            if self._game[x][y] == "_": # Check if tile is empty
                self._game[x][y] = 'x'
                break

    def player2_input(self):
        while True:
            x = random.randint(0, 2)
            y = random.randint(0, 2)
            if self._game[x][y] == "_": # Check if tile is empty
                self._game[x][y] = 'o'
                break
